core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -734,6 +734,31 @@ private QueryModel bubbleUpOrderByAndLimitFromUnion(QueryModel model) throws Sql
         return model;
     }
 
+    //pushing predicates to sample by model is only allowed for sample by fill none align to calendar and expressions on non-timestamp columns
+    //pushing for other fill options or sample by first observation could alter result 
+    private boolean canPushToSampleBy(final QueryModel model, ObjList<CharSequence> expressionColumns) {
+        ObjList<ExpressionNode> fill = model.getSampleByFill();
+        int fillCount = fill.size();
+        boolean isFillNone = fillCount == 0 || (fillCount == 1 && SqlKeywords.isNoneKeyword(fill.getQuick(0).token));
+
+        if (!isFillNone || model.getSampleByOffset() == null) {
+            return false;
+        }
+
+        CharSequence timestamp = findTimestamp(model);
+        if (timestamp == null) {
+            return true;
+        }
+
+        for (int i = 0, n = expressionColumns.size(); i < n; i++) {
+            if (Chars.equalsIgnoreCase(expressionColumns.get(i), timestamp)) {
```

Comment:
Will this catch something like 

```
select * from (
  select timestamp, symbol, side, first(price)
  from trades
  SAMPLE by 1d align to calendar
)
where symbol = 'BTC-USD' and side = 'sell' and dateadd('h', -1, timestamp) > '2022-12'
```

and prevent pushing down the timestamp filter?

This implementation is also is all or nothing which is not the best. In reality with the above query it's ok to push down `symbol` and `side` filters and leave timestamp filter externally.

Reply:
Please have a look at https://github.com/questdb/questdb/pull/2945/files#diff-5f98c23781bda3e3b7422770c5e1ef57d48e6ef6de4594abe849fea648d701e9 . 
This implementation is per where clause expression so it does push non-ts expressions if it makes sense .